tests/subsys/dfu/dfu_multi_image/src/main.c

```diff
@@ -0,0 +1,270 @@
+#include <sys/util.h>
+#include <ztest.h>
+
+#include <string.h>
+#include <stdint.h>
+#include <stddef.h>
+
+/*
+ * Implement image writer which verifies written data instead of actually saving it.
+ */
+
+struct expected_image {
+	int image_id;
+	/* expected image content */
+	const char *content;
+	size_t content_size;
+};
+
+struct expected {
+	/* images expected to be written when unpacking a given update package */
+	struct expected_image images[CONFIG_DFU_MULTI_IMAGE_MAX_IMAGE_COUNT];
+	size_t image_count;
+};
+
+struct comparison_context {
+	struct expected expected;
+	size_t current_image_no;
+	size_t current_image_offset;
+};
+
+static struct comparison_context ctx;
+
+static int image_comparator_open(int image_id)
+{
+	zassert_true(ctx.current_image_no < ctx.expected.image_count, "Too many images written");
+	zassert_equal(ctx.expected.images[ctx.current_image_no].image_id, image_id,
+		      "Unexpected image id");
+	zassert_equal(ctx.current_image_offset, 0, "Opening image while already in progress");
+
+	return 0;
+}
+
+static int image_comparator_write(const uint8_t *chunk, size_t chunk_size)
+{
+	const struct expected_image *image = &ctx.expected.images[ctx.current_image_no];
+
+	zassert_true(ctx.current_image_offset + chunk_size <= image->content_size,
+		     "Too large image written");
+	zassert_ok(memcmp(image->content + ctx.current_image_offset, chunk, chunk_size),
+		   "Unexpected image content");
+
+	ctx.current_image_offset += chunk_size;
+
+	return 0;
+}
+
+static int image_comparator_close(bool success)
+{
+	const struct expected_image *image = &ctx.expected.images[ctx.current_image_no];
+
+	zassert_true(success, "Closing image with failure");
+	zassert_equal(ctx.current_image_offset, image->content_size, "Too small image written");
+
+	ctx.current_image_no++;
+	ctx.current_image_offset = 0;
+
+	return 0;
+}
+
+/*
+ * Generic test that verifies expected image data is written when processing a DFU package.
+ */
+
+static int comparison_test(const uint8_t *package, size_t package_size,
+			   const struct expected *expected, uint8_t *buffer, size_t buffer_size,
+			   size_t chunk_size)
+{
+	int err;
+
+	ctx.expected = *expected;
+	ctx.current_image_no = 0;
+	ctx.current_image_offset = 0;
+
+	/* Initialize DFU and register image writers for all expected images */
+
+	err = dfu_multi_image_init(buffer, buffer_size, NULL);
+
+	if (err) {
+		return err;
+	}
+
+	for (size_t i = 0; i < expected->image_count; ++i) {
+		struct dfu_image_writer writer = { .image_id = expected->images[i].image_id,
+						   .open = image_comparator_open,
+						   .write = image_comparator_write,
+						   .close = image_comparator_close };
+
+		err = dfu_multi_image_register_writer(&writer);
+
+		if (err) {
+			return err;
+		}
+	}
+
+	/* Write the package using the requested chunk size */
+
+	for (size_t i = 0; i < package_size; i += chunk_size) {
+		err = dfu_multi_image_write(i, package + i, MIN(chunk_size, package_size));
```

Comment:
So all tests base on this helper function, and every test uses dfu_multi_image_write with consecutive writes. But the interface of this function (because of presence of  offset), does not enforce consecutive write.

We should also test for scenario like this:
dfu_multi_image_write(0, buff, 100);
dfu_multi_image_write(100, buff, 100);
dfu_multi_image_write(300, buff, 100); // leave a gap for 100 Bytes
dfu_multi_image_write(200, buff, 100); // fill the gap

This could occur if, for example, the packets were rearranged in transit / router.  From the interface point of view, this is not prohibited, so we should also test this scenario.